tensorflow/core/kernels/while_op_test.cc

```diff
@@ -0,0 +1,168 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0(the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include "tensorflow/c/experimental/stream_executor/stream_executor.h"
+
+#include "tensorflow/c/experimental/stream_executor/stream_executor_internal.h"
+#include "tensorflow/c/experimental/stream_executor/stream_executor_test_util.h"
+#include "tensorflow/cc/client/client_session.h"
+#include "tensorflow/cc/framework/scope.h"
+#include "tensorflow/cc/ops/standard_ops.h"
+#include "tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.h"
+#include "tensorflow/core/framework/function_testlib.h"
+#include "tensorflow/core/graph/node_builder.h"
+#include "tensorflow/core/kernels/ops_testutil.h"
+#include "tensorflow/stream_executor/event.h"
+#include "tensorflow/stream_executor/multi_platform_manager.h"
+
+namespace stream_executor {
+namespace {
+
+class WhileOpTest : public tensorflow::OpsTestBase {
+ protected:
+  WhileOpTest() {}
+  void SetUp() override {
+    test_util::PopulateDefaultPlatform(&platform_, &platform_fns_);
+    test_util::PopulateDefaultDeviceFns(&device_fns_);
+    test_util::PopulateDefaultStreamExecutor(&se_);
+    test_util::PopulateDefaultTimerFns(&timer_fns_);
+  }
+  void TearDown() override {}
+
+  SP_Platform platform_;
+  SP_PlatformFns platform_fns_;
+  SP_DeviceFns device_fns_;
+  SP_StreamExecutor se_;
+  SP_TimerFns timer_fns_;
+};
+
+TEST_F(WhileOpTest, WhileOpCPUBuildWithPluggableDevice) {
+  using namespace tensorflow;
+
+  const std::string platform_name = "MY_TEST";
+  const std::string platform_type = "FAKE";
+  platform_.name = platform_name.c_str();
+  platform_.type = platform_type.c_str();
+
+  static bool memcpy_d2h_called = false;
+  se_.memcpy_dtoh = [](const SP_Device* device, SP_Stream stream,
```

Comment:
I was thinking that we could put all these stream executor functions into the mock pluggable device in [test_pluggable_device.cc](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/experimental/stream_executor/test/test_pluggable_device.cc). This way other tests can just link to [test_pluggable_device.so](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/experimental/stream_executor/test/BUILD;l=13;drc=e846e2ceda6a2bff0f0f25494cf0af12555b4835) without having to implement all these functions from scratch every time. I'm okay with them being here now since we are close to the 2.7 branch cut. (I'm sorry I couldn't get to this PR sooner!) I'll refactor them out later.